File: training-data/swift/kruskal_mst.swift
final class UnionFind {
    private var parent: [Int]

    init(_ n: Int) {
        parent = Array(0..<n)
    }

    func find(_ x: Int) -> Int {
        if parent[x] != x { parent[x] = find(parent[x]) }
        return parent[x]
    }

    func union(_ a: Int, _ b: Int) -> Bool {
        let rootA = find(a), rootB = find(b)
        if rootA == rootB { return false }
        parent[rootA] = rootB
        return true
    }
}

func kruskalMST(nodeCount: Int, edges: [(Int, Int, Int)]) -> [(Int, Int, Int)] {
    let uf = UnionFind(nodeCount)
    var mst: [(Int, Int, Int)] = []
    for edge in edges.sorted(by: { $0.2 < $1.2 }) {
        if uf.union(edge.0, edge.1) {
            mst.append(edge)
        }
    }
    return mst
}

let edges = [
    (0, 1, 4), (0, 2, 1), (1, 2, 2),
    (1, 3, 5), (2, 3, 8), (3, 4, 3),
]

let mst = kruskalMST(nodeCount: 5, edges: edges)
for (u, v, w) in mst {
    print("\(u)-\(v): \(w)")
}
print("total weight:", mst.reduce(0) { $0 + $1.2 })

File: training-data/scala/KruskalMST.scala
class DisjointSet(n: Int) {
  private val parent = Array.tabulate(n)(identity)

  def find(x: Int): Int = {
    if (parent(x) != x) parent(x) = find(parent(x))
    parent(x)
  }

  def union(a: Int, b: Int): Boolean = {
    val rootA = find(a)
    val rootB = find(b)
    if (rootA == rootB) false
    else {
      parent(rootA) = rootB
      true
    }
  }
}

object KruskalMST {
  case class Edge(u: Int, v: Int, weight: Int)

  def mst(n: Int, edges: List[Edge]): (List[Edge], Int) = {
    val dsu = new DisjointSet(n)
    val sorted = edges.sortBy(_.weight)
    val chosen = sorted.filter(e => dsu.union(e.u, e.v))
    (chosen, chosen.map(_.weight).sum)
  }

  def main(args: Array[String]): Unit = {
    val edges = List(
      Edge(0, 1, 4), Edge(0, 2, 1), Edge(2, 1, 2),
      Edge(1, 3, 5), Edge(2, 3, 8), Edge(3, 4, 3)
    )
    val (chosen, total) = mst(5, edges)
    chosen.foreach(e => println(s"${e.u} - ${e.v} : ${e.weight}"))
    println(s"total weight: $total")
  }
}
